Caching Doc: Create separate entry for opcache (#3819)
owncloud/docs: modules/admin_manual/pages/configuration/server/caching_configuration.adoc

A PHP opcode Cache (OPcache)::
An opcode cache stores compiled PHP scripts so they don’t need to be re-compiled every time they are called. These compiled PHP scripts are stored in-memory, on the server on which they’re compiled.

A Data Cache::
A data cache stores copies of _data_, _templates_, and other types of _information-based files_. Depending on the cache implementation, it can be either _local_, or specific, to one server, or _distributed_ across multiple servers. This cache type is ideal when you have a scale-out installation.

In addition, we suggest to use *External Transactional File Locking* which reduces load on the database significantly.

== Supported Caching Backends

The caching backends supported by ownCloud are:

* xref:opcache[Opcache] +
  This is an opcode cache only and does *not* cache any data.
  Opcache is bundled with PHP from version 5.5.0 and later.
* xref:apcu[APCu] +
  This is a data cache only and does *not* cache any opcode.
  APCu 4.0.6 and up is required.
* xref:redis[Redis] +

>   This is an in-memory data structure store (cache) for single and multi-server ownCloud installations, provides file locking and can be setup in local or distributed environments. Consider Redis younger, more feature-rich and configurable than memchached. At minimum, avrsion 2.2.6 or higher of the PHP Redis extension is required.

Reviewer: ```suggestion
  This is an in-memory data structure store (cache) for single and multi-server ownCloud installations, provides file locking and can be setup in local or distributed environments. Consider Redis younger, more feature-rich and configurable than memcached. At minimum, version 2.2.6 or higher of the PHP Redis extension is required.
```

* xref:memcached[Memcached] +
  This is a distributed cache for multi-server ownCloud installations and has *no* file locking capabilities.

> See following page to learn more about the {redis-memcached-url}[Redis vs. Memcached – 2021 Comparison].

Reviewer: ```suggestion
See the following page to learn more about the {redis-memcached-url}[Redis vs. Memcached – 2021 Comparison].
```

* xref:memcached[Memcached] +
  This is a distributed cache for multi-server ownCloud installations and has *no* file locking capabilities.

See the following page to learn more about the {redis-memcached-url}[Redis vs. Memcached – 2021 Comparison].

[NOTE]
====
You may use _both_ a local and a distributed cache. The recommended ownCloud caches are APCu and Redis. If you do not install and enable a local memory cache you will see a warning on your ownCloud admin page. If you enable only a distributed cache in your `config.php` (`memcache.distributed`) and not a local cache (`memcache.local`) you will still see the cache warning.
====

=== Cache Directory Location

The cache directory defaults to `data/$user/cache` where `$user` is the current user. You may use the `'cache_path'` directive in your configuration for different locations. For details see the  xref:configuration/server/config_sample_php_parameters.adoc#define-the-location-of-the-cache-folder[Define the location of the cache folder] description.

== Cache Types

=== Opcache

Opcache should be enabled by default in your php installation. To check it, run the following command:

[source,php]
----

> php -r 'phpinfo();' | grep opcache.enable

Author: Slight caveat that opcache may be enabled on webserver and not on cli

The installer will automatically launch Redis and configure it to launch at startup.

After that, assuming that you don’t encounter any errors, restart Apache and the extension is ready to use.

==== Additional notes for Redis vs. APCu on Memory Caching

APCu is faster at local caching than Redis. If you have enough memory, use APCu for memory caching and Redis for file locking. If you are low on memory, use Redis for both.

==== Clearing the Redis Cache

The Redis cache can be flushed from the command-line using the {rediscli_url}[redis-cli tool], as in the following example:

----
sudo redis-cli
SELECT <dbIndex>
FLUSHDB
----

> `<dbIndex>` is the number of Redis database where the cache is stored. It is zero by default at ownCloud. To check what yours is currently set to, check the `dbindex` value in `config/config.php`. Here’s an example of what to look for:

Reviewer: Where is the example?

Reviewer 2: Thanks for spotting it 👍 
These things always happen on a bigger review of existing documents...

Reviewer 2: fixed

==== Clearing the Memcached Cache

> The Memcached cache can be flushed from the command-line using a range of common Linux/Unix tools, including `netcat` and `telnet`. The following example uses telnet to login, run the {flushall_url}[flush_all command], and logout:

Reviewer 3: ```suggestion
The Memcached cache can be flushed from the command line, using a range of common Linux/Unix tools including `netcat` and `telnet`. The following example uses telnet to log in, run the {flushall_url}[flush_all command], and log out:
```

==== Redis Configuration Using Unix Sockets

> If Redis is running on the same server as ownCloud, it is recommended to configure it to use Unix sockets. Then, configure ownCloud to communicate with Redis, as in the following example. 

Reviewer 3: ```suggestion
If Redis is running on the same server as ownCloud, it is recommended to configure it to use Unix sockets. Then, configure ownCloud to communicate with Redis as in the following example. 
```

 ],
----

CAUTION: For enhanced security, it is recommended to configure Redis to require a password. See {redis_security_url} for more information.

== Caching Exceptions

> If ownCloud is configured to use either Memcached or Redis as a memory cache, please be aware that you may encounter issues with functionality. When these occur, it is usually a result of PHP being incorrectly configured, or the relevant PHP extension not being available.

Reviewer 3: ```suggestion
If ownCloud is configured to use either Memcached or Redis as a memory cache, you may encounter issues with functionality. When these occur, it is usually a result of PHP being incorrectly configured or the relevant PHP extension not being available.
```
